File: leetcode-cn/18.四数之和.py
#
# @lc app=leetcode.cn id=18 lang=python3
#
# [18] 四数之和
#

# @lc code=start
from typing import List


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        ans = []
        nums.sort()
        n = len(nums)

        for i in range(n - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            if nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3] > target:
                break
            if nums[i] + nums[n - 1] + nums[n - 2] + nums[n - 3] < target:
                continue

            for j in range(i + 1, n - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue

                if nums[i] + nums[j] + nums[j + 1] + nums[j + 2] > target:
                    break
                if nums[i] + nums[j] + nums[n - 1] + nums[n - 2] < target:
                    continue

                left = j + 1
                right = len(nums) - 1
                tt = target - nums[i] - nums[j]

                while left < right:
                    if nums[left] + nums[right] == tt:
                        ans.append([nums[i], nums[j], nums[left], nums[right]])

                        while left < right and nums[left] == nums[left + 1]:
                            left += 1
                        left += 1

                        while left < right and nums[right] == nums[right - 1]:
                            right -= 1
                        right -= 1

                    elif nums[left] + nums[right] < tt:
                        left += 1
                    elif nums[left] + nums[right] > tt:
                        right -= 1
        return ans


# @lc code=end

File: leetcode-cn/C++/18.四数之和.cpp
/*
 * @lc app=leetcode.cn id=18 lang=cpp
 *
 * [18] 四数之和
 */

// @lc code=start
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<vector<int>> fourSum(vector<int> &nums, int target)
    {
        sort(nums.begin(), nums.end());
        int n = nums.size();

        vector<vector<int>> ans;
        long long sum_ = 0;
        for (int first = 0; first < n - 3; first++)
        {
            if (first > 0 && nums[first] == nums[first - 1])
            {
                continue;
            }

            sum_ = nums[first] + nums[first + 1] + nums[first + 2] + nums[first + 3];
            if (sum_ > target)
            {
                break;
            }

            for (int second = first + 1; second < n; second++)
            {
                if (second > first + 1 && nums[second - 1] == nums[second])
                {
                    continue;
                }

                int third = second + 1;
                int forth = n - 1;

                while (third < forth)
                {
                    long long sum_ = 0;
                    for (int i : {first, second, third, forth})
                    {
                        sum_ += nums[i];
                    }

                    if (sum_ < target)
                    {
                        third++;
                    }
                    else if (sum_ > target)
                    {
                        forth--;
                    }
                    else
                    {
                        ans.push_back({nums[first], nums[second], nums[third], nums[forth]});

                        while (third < forth && nums[third] == nums[third + 1])
                        {
                            third++;
                        }
                        third++;

                        while (third < forth && nums[forth] == nums[forth - 1])
                        {
                            forth--;
                        }
                        forth--;
                    }
                }
            }
        }

        return ans;
    }
};
// @lc code=end
